pkg/typesystem/typesystem_test.go

```diff
@@ -434,6 +434,125 @@ func TestInvalidRewriteValidations(t *testing.T) {
 			},
 			err: ErrRelationUndefined,
 		},
+		{
+			name: "Fails_If_Using_This_As_Relation_Name",
+			model: &openfgapb.AuthorizationModel{
+				SchemaVersion: SchemaVersion1_0,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "repo",
+						Relations: map[string]*openfgapb.Userset{
+							"this": This(),
+						},
+					},
+				},
+			},
+			err: ErrReservedKeywords,
+		},
+		{
+			name: "Fails_If_Using_Self_As_Relation_Name",
+			model: &openfgapb.AuthorizationModel{
+				SchemaVersion: SchemaVersion1_0,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "repo",
+						Relations: map[string]*openfgapb.Userset{
+							"self": This(),
+						},
+					},
+				},
+			},
+			err: ErrReservedKeywords,
+		},
+		{
+			name: "Fails_If_Using_This_As_Type_Name",
+			model: &openfgapb.AuthorizationModel{
+				SchemaVersion: SchemaVersion1_0,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "this",
+						Relations: map[string]*openfgapb.Userset{
+							"viewer": This(),
+						},
+					},
+				},
+			},
+			err: ErrReservedKeywords,
+		},
+		{
+			name: "Fails_If_Using_Self_As_Type_Name",
+			model: &openfgapb.AuthorizationModel{
+				SchemaVersion: SchemaVersion1_0,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "self",
+						Relations: map[string]*openfgapb.Userset{
+							"viewer": This(),
+						},
+					},
+				},
+			},
+			err: ErrReservedKeywords,
+		},
+		{
+			name: "Fails_If_Auth_Model_1.1_Has_A_Cycle",
```

Comment:
Can this test be moved to `tests/model`? Probably some of the others can't as they will fail parsing before we even make it to the write part, but the ones that can would be nice to move.

Reply:
😱 i had no idea those existed. will check it out